File: sdk/nodejs/primus-identity-validator/src/services/openIdConfigurationService.ts
import axios, { AxiosInstance } from 'axios';
import { OpenIdConfiguration } from '../types';

interface CachedConfiguration {
  configuration: OpenIdConfiguration;
  cachedAt: Date;
  expiresAt: Date;
}

/**
 * Service for fetching and caching OpenID Connect configuration (Azure AD and generic OIDC)
 */
export class OpenIdConfigurationService {
  private readonly httpClient: AxiosInstance;
  private readonly configCache: Map<string, CachedConfiguration>;
  private readonly cacheTtl: number; // milliseconds
  private readonly fetchLock: Map<string, Promise<OpenIdConfiguration>>;

  constructor(cacheTtl: number = 24 * 60 * 60 * 1000) {
    this.httpClient = axios.create({ timeout: 10000 });
    this.configCache = new Map();
    this.cacheTtl = cacheTtl;
    this.fetchLock = new Map();
  }

  /**
   * Gets OpenID Connect configuration for a given authority or Azure AD tenant ID
   * @param authorityOrTenant Authority URL (https://...) or tenant ID (Azure AD)
   * @returns OpenID Connect configuration
   */
  async getConfigurationAsync(authorityOrTenant: string): Promise<OpenIdConfiguration> {
    const wellKnownUrl = this.getWellKnownUrl(authorityOrTenant);
    
    // Check cache
    const cached = this.configCache.get(wellKnownUrl);
    if (cached && Date.now() < cached.expiresAt.getTime()) {
      return cached.configuration;
    }
    
    // Check if fetch is in progress
    const inProgress = this.fetchLock.get(wellKnownUrl);
    if (inProgress) {
      return inProgress;
    }
    
    // Fetch from Azure AD
    const fetchPromise = this.fetchConfiguration(wellKnownUrl);
    this.fetchLock.set(wellKnownUrl, fetchPromise);
    
    try {
      const config = await fetchPromise;
      
      // Cache the configuration
      this.configCache.set(wellKnownUrl, {
        configuration: config,
        cachedAt: new Date(),
        expiresAt: new Date(Date.now() + this.cacheTtl)
      });
      
      return config;
    } finally {
      this.fetchLock.delete(wellKnownUrl);
    }
  }

  /**
   * Fetches OpenID Connect configuration from Azure AD
   * @param url Well-known configuration URL
   * @returns OpenID Connect configuration
   */
  private async fetchConfiguration(url: string): Promise<OpenIdConfiguration> {
    const response = await this.httpClient.get<OpenIdConfiguration>(url);
    return response.data;
  }

  /**
   * Gets the well-known OpenID Connect configuration URL
   * @param authorityOrTenant Authority URL or Azure AD tenant ID
   * @returns Well-known configuration URL
   */
  private getWellKnownUrl(authorityOrTenant: string): string {
    // If caller passed a bare tenant ID (no scheme), assume Azure AD
    const isBareTenantId = !authorityOrTenant.includes('://');
    const authority = isBareTenantId
      ? `https://login.microsoftonline.com/${authorityOrTenant}`
      : authorityOrTenant;

    const url = new URL(authority);
    const isAzureHost =
      url.hostname.toLowerCase().includes('login.microsoftonline.com') ||
      url.hostname.toLowerCase().includes('sts.windows.net');

    // Append /v2.0 for Azure AD when missing; leave others untouched
    let path = url.pathname.endsWith('/') ? url.pathname.slice(0, -1) : url.pathname;
    if (isAzureHost && !path.toLowerCase().endsWith('/v2.0')) {
      path = `${path}/v2.0`;
    }

    // Build well-known path
    url.pathname = `${path}/.well-known/openid-configuration`;
    url.search = '';

    return url.toString();
  }

  /**
   * Clears the configuration cache
   */
  clearCache(): void {
    this.configCache.clear();
  }

  /**
   * Gets the number of cached configurations
   */
  get cacheCount(): number {
    return this.configCache.size;
  }
}

File: sdk/dotnet/PrimusSaaS.Identity.Validator/Services/OpenIdConfigurationService.cs
using PrimusSaaS.Identity.Validator.Models;
using System.Text.Json;

namespace PrimusSaaS.Identity.Validator.Services;

/// <summary>
/// Service for fetching and caching OpenID Connect configuration.
/// </summary>
public class OpenIdConfigurationService
{
    private readonly HttpClient _httpClient;
    private readonly Dictionary<string, CachedConfiguration> _configCache = new();
    private readonly TimeSpan _cacheTtl;
    private readonly SemaphoreSlim _fetchLock = new(1, 1);

    /// <summary>
    /// Initializes a new instance of the <see cref="OpenIdConfigurationService"/> class.
    /// </summary>
    /// <param name="httpClient">HTTP client for fetching configuration.</param>
    /// <param name="cacheTtl">Cache TTL for configuration. Defaults to 24 hours.</param>
    public OpenIdConfigurationService(HttpClient? httpClient = null, TimeSpan? cacheTtl = null)
    {
        _httpClient = httpClient ?? new HttpClient();
        _cacheTtl = cacheTtl ?? TimeSpan.FromHours(24);
    }

    /// <summary>
    /// Gets the OpenID Connect configuration for an Azure AD tenant.
    /// </summary>
    /// <param name="tenantId">The Azure AD tenant ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The OpenID configuration.</returns>
    public async Task<PrimusOpenIdConfiguration> GetConfigurationAsync(
        string tenantId,
        CancellationToken cancellationToken = default)
    {
        var wellKnownUrl = GetWellKnownUrl(tenantId);

        // Check cache first
        if (_configCache.TryGetValue(wellKnownUrl, out var cached))
        {
            if (DateTimeOffset.UtcNow < cached.ExpiresAt)
            {
                return cached.Configuration;
            }

            // Expired - remove from cache
            _configCache.Remove(wellKnownUrl);
        }

        // Fetch from Azure AD
        await _fetchLock.WaitAsync(cancellationToken);
        try
        {
            // Double-check after acquiring lock
            if (_configCache.TryGetValue(wellKnownUrl, out cached))
            {
                if (DateTimeOffset.UtcNow < cached.ExpiresAt)
                {
                    return cached.Configuration;
                }
            }

            var response = await _httpClient.GetAsync(wellKnownUrl, cancellationToken);
            response.EnsureSuccessStatusCode();

            var json = await response.Content.ReadAsStringAsync(cancellationToken);
            var configuration = JsonSerializer.Deserialize<PrimusOpenIdConfiguration>(json);

            if (configuration == null)
            {
                throw new InvalidOperationException("Failed to deserialize OpenID configuration.");
            }

            // Cache the configuration
            _configCache[wellKnownUrl] = new CachedConfiguration
            {
                Configuration = configuration,
                CachedAt = DateTimeOffset.UtcNow,
                ExpiresAt = DateTimeOffset.UtcNow.Add(_cacheTtl)
            };

            return configuration;
        }
        finally
        {
            _fetchLock.Release();
        }
    }

    /// <summary>
    /// Gets the well-known configuration URL for an Azure AD tenant.
    /// </summary>
    /// <param name="tenantId">The Azure AD tenant ID.</param>
    /// <returns>The well-known URL.</returns>
    public static string GetWellKnownUrl(string tenantId)
    {
        return $"https://login.microsoftonline.com/{tenantId}/v2.0/.well-known/openid-configuration";
    }

    /// <summary>
    /// Clears the configuration cache.
    /// </summary>
    public void ClearCache()
    {
        _configCache.Clear();
    }

    private class CachedConfiguration
    {
        public PrimusOpenIdConfiguration Configuration { get; set; } = null!;
        public DateTimeOffset CachedAt { get; set; }
        public DateTimeOffset ExpiresAt { get; set; }
    }
}
